### PR TITLE
Implement best matches only option with browser integration

### LINKED ISSUE #1112: KeepassXC - HTTP: does not respect the BestMatchOnly setting
Currently the setting is a noop

https://github.com/keepassxreboot/keepassxc/blob/139658b5c33a39d261a5f3b0ba89c6b93a4a38f9/src/http/HttpSettings.h#L32

## Expected Behavior
Setting uses a deterministic sort on the returned list by best match, only returning the 'top' best match.

## Current Behavior
Setting is a noop, doing nothing

## Possible Solution
Implement the sort function

## Steps to Reproduce (for bugs)
1. enable BestMatchOnly in the HTTP settings
2. Run a browser plugin (KeePassHttp-Connector on Firefox 57 for me).
3. Try to log into a site with that has multiple entries in KeepassXC (different ports in my case)
4. See that multiple entries are returned

## Context
Finish the http plugin

## Debug Info
Version Built/tested: 4c2bc56d2ea89d4c36b378082de15964a9355741

Operating system: Gentoo
CPU architecture: amd64
Kernel: 4.9.58

Enabled extensions:
- http

### PR DESCRIPTION
**Pull request by varjolintu**
Returning only the best matched entries option hasn't worked properly. This fixes it and only returns entries with sort priority 100.

## Description
It's not perfect but a small improvement. If you can think anything that could improve this PR please let me know.

## Motivation and context
Fixes 
https://github.com/keepassxreboot/keepassxc-browser/issues/84

Fixes #1112

## How has this been tested?
Manually.

## Types of changes
- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by rugk** [comment]
Better add "fixes" or so, before each issue you link, so it is [automatically closed](https://help.github.com/articles/closing-issues-via-commit-messages/) when a PR is merged. (I know, third time I comment that. :smile:)

**Comment by rugk** [comment]
BTW as for me I could not even find that setting in the UI in KeePassXC. The description here sounds as if the option is there, but just did not work. Or did you hide it deliberately, because you needed to make it work?

**Comment by varjolintu** [comment]
The setting has been hidden since the new browser integration was merged because it was not implemented.

File: src/browser/BrowserService.cpp [hunk 1]
```diff
@@ -513,8 +513,16 @@ QList<Entry*> BrowserService::sortEntries(QList<Entry*>& pwEntries, const QStrin
     const QString baseSubmitUrl = url.toString(QUrl::StripTrailingSlash | QUrl::RemovePath | QUrl::RemoveQuery | QUrl::RemoveFragment);
 
     QMultiMap<int, const Entry*> priorities;
-    for (const Entry* entry : pwEntries) {
-        priorities.insert(sortPriority(entry, host, submitUrl, baseSubmitUrl), entry);
+    for (Entry* entry : pwEntries) {
+        const int priority = sortPriority(entry, host, submitUrl, baseSubmitUrl);
+
+        // Ignore matches that aren't exact enough
+        if (BrowserSettings::bestMatchOnly() && priority < 100) {
```
Comment (droidmonkey) [inline src/browser/BrowserService.cpp]: Does 100 represent percent matching? If not, this might need some explaining.

Comment (varjolintu) [reply src/browser/BrowserService.cpp]: Not exactly. It's from the old HTTP implementation that tries to mimic a percentage. See https://github.com/varjolintu/keepassxc/blob/e0a06bd515d77a9b62991a5ee2a8a4740cf79bb3/src/browser/BrowserService.cpp#L650

Comment (droidmonkey) [reply src/browser/BrowserService.cpp]: Interesting... This will be a target for 2.4 haha

**Comment by varjolintu** [comment]
Rebased. I think this is quite ready for merging.

**Comment by droidmonkey** [comment]
I fixed an error where zero priority matches were being filtered by the sort function. We should really collapse a lot of the filtering and sorting happening in the browser service.

**Pull request merged**